### PR TITLE
Dev/ft/sub module logger

### PR DESCRIPTION
**Pull request by MathieuCassagne**

### CODE REVIEW

File: lib/Logger.js [hunk 1]
```diff
@@ -30,14 +30,19 @@ class Logger {
...
+     *                                 log entries.
+     *
      * @see [Bunyan's documentation]{@link
      * https://github.com/trentm/node-bunyan/blob/master/README.md#streams} for
      * a more detailed description of the streams array configuration.
      *
      * @returns {undefined}
      */
-    constructor(name, config) {
+    constructor(name, config, subModuleName) {
```
Comment (DavidPineauScality) [inline lib/Logger.js]: I'd have put subModule name as the second argument for consistency,  but doing it this way keeps the backwards compatibility so :+1:

File: lib/Logger.js [hunk 2]
```diff
@@ -49,6 +54,40 @@ class Logger {
...
+
+    /**
+     *  This function creates a new Logger, from another one,
+     *  changing its submodule field.
+     *
+     *  @param {Logger} baseLogger - Base Logger to create a new one.
+     *  @param {String} submodule - The name of the submodule.
+     *  @return {Logger} - The new logger with the new submodule.
+     */
+    static updateSubmodule(baseLogger, submodule) {
```
Comment (DavidPineauScality) [inline lib/Logger.js]: Few comments here:
1. What is the point of updating the submodule for a "module logger" ?
2. Why create a new logger in an "updateSubmodule" ? Maybe the function needs renaming
3. Should we not only use that approach for the requestLogger, as previously discussed ?

File: lib/Logger.js [hunk 3]
```diff
@@ -49,6 +54,40 @@ class Logger {
...
+    }
+
+    /**
+     *  This factory creates a function which update the submodule of a
+     *  RequestLogger according to a ModuleLogger submodule.
+     *
+     *  @param {Logger} log - Base Logger to create a new one.
+     *  @return {RequestLogger} - The new logger with the new submodule.
+     */
+    static updateRequestLoggerSubmoduleFactory(log) {
```
Comment (DavidPineauScality) [inline lib/Logger.js]: Nope, I'm against the Java-isation with meaningless factories and very long names
This should be a method of a requestLogger instance instead, as we discussed.

File: tests/unit/Logger.js [hunk 4]
```diff
@@ -8,8 +8,8 @@ const loggingMisuseGenerator = Utils.loggingMisuseGenerator;
 const DummyLogger = Utils.DummyLogger;
 
 const Config = require('../../lib/Config.js');
+const Logger = require('../../lib/Logger.js');
 const RequestLogger = require('../../lib/RequestLogger.js');
-const Logger = require('../../index.js').Logger;
```
Comment (DavidPineauScality) [inline tests/unit/Logger.js]: why this change ?

File: lib/Logger.js (hunk 2 above)
Comment (MathieuCassagne) [inline lib/Logger.js]: I planned to pass the Logger from submodule to submodule, but your comments make me realize it's the same things as creating a new one in each submodules ...
So you're right, it's quite useless.

File: tests/unit/Logger.js (hunk 4 above)
Comment (MathieuCassagne) [inline tests/unit/Logger.js]: I first move it for the alphabetical order, and then I wanted to make it looks like others.

File: lib/RequestLogger.js [hunk 5]
```diff
@@ -222,6 +222,19 @@ class RequestLogger {
...
     /**
+     *  This function creates a new RequestLogger object from a ModuleLogger
+     *  already existing in order to have the submodule field from the
+     *  ModuleLogger.
+     *
+     *	@param {ModuleLogger} ml - The base ModuleLogger
+     *	@return {RequestLogger} - The new RequestLogger created
+     *				  with the ModuleLogger
+     */
+    scope(ml) {
```
Comment (DavidPineauScality) [inline lib/RequestLogger.js]: Afair, the API we discussed was more along the lines of:
```
scope(subname) {
    return ScopedRequestLogger(this, subname);
}
```
with the ScopedRequestLogger  being very similar to the EndLogger in terms of API and capacity.

Why didn't you go that way ? (I'm feeling like you are attempting various things without attempting what we discussed. Is there any reason?)

File: tests/unit/Logger.js (hunk 4 above)
Comment (DavidPineauScality) [inline tests/unit/Logger.js]: Mh.... okay, I feel like it was not necessary but so be it.

File: lib/RequestLogger.js (hunk 5 above)
Comment (MathieuCassagne) [inline lib/RequestLogger.js]: Because in fact the ScopedRequestLogger is exactly the same as a RequestLogger with only the submodule differing, but present in all cases, so the ScopedRequestLogger class will just be a kind of copy/paste of RequestLogger. So I thought creating a new RequestLogger using the right ModuleLogger was enough and simpler.

Comment (DavidPineauScality) [inline lib/RequestLogger.js]: Creating a new RequestLogger effectively changes the request UID. That's why this is out of the question, messing with our request tracking capabilities. Hence my suggestion to _wrap_ a requestLogger, much like the EndLogger. You  do not need to re-write the RequestLogger; only to wrap it and add what you need.

**Comment by DavidPineauScality** [comment]
After offline discussion with @alexandre-merle, we improved the design originally discussed with you.
To make things a bit faster, I enjoyed myself a bit and implemented the new design in PR #86.

I'll let you do what's needed with this PR, but it should be rendered obsolete by the new design I offer in the PR I opened.

**Comment by MathieuCassagne** [comment]
@DavidPineauScality: I think the easier is to take your work, since this one is used nowhere.

Closing this PR in favor of #86.
